### PR TITLE
Task 1 - class store() implementation

### LINKED ISSUE #39: Task 1 - class Store
Napisz klasę Store, która będzie umożliwiała dokonywanie zakupów. Wykorzystaj poniższy enum i funkcje.

```
enum class Response {done, lack_of_money, lack_of_cargo, lack_of_space};

Response buy(Cargo* cargo, size_t amount, Player* player);
Response sell(Cargo* cargo, size_t amount, Player* player);

```

### PR DESCRIPTION
**Pull request by Petrus9107**
Sharing Task_1 code, for further implementations some remarks still pending (code formatting, my own review etc.)

### CODE REVIEW

File: shm/src/ship.cpp [hunk 1]
```diff
@@ -41,18 +41,37 @@ size_t Ship::getSpeed() const     { return speed_; }
...
+
 size_t Ship::getAvailableSpace() const {
-    size_t occupied = std::accumulate(cargo_.begin(),
-                                      cargo_.end(),
-                                      0,
-                                      [](size_t sum, const auto& cargo) {
-                                          return sum += cargo->getAmount();
-                                      });
-    return capacity_ - occupied;
+    Common::getAvailableSpace(capacity_, cargo_);
```
Comment (mariuszlisowski) [inline shm/src/ship.cpp]: return Common::getAvailableSpace(capacity_, cargo_);

File: shm/src/ship.cpp [hunk 2]
```diff
@@ -41,18 +41,37 @@ size_t Ship::getSpeed() const     { return speed_; }
...
-        return nullptr;
-    }
-    return cargo_[index];
+
+
+Cargo* Ship::getCargo(const std::string& name) {
+    return Common::getCargo(name, cargo_);
 }
+
+
```
Comment (mariuszlisowski) [inline shm/src/ship.cpp]: one newline redundant

File: shm/src/ship.cpp [hunk 3]
```diff
@@ -41,18 +41,37 @@ size_t Ship::getSpeed() const     { return speed_; }
 std::string Ship::getName() const { return name_; }
 size_t Ship::getId() const        { return id_; }
 
-std::shared_ptr<Cargo> Ship::getCargo(const size_t index) {
-    if (index >= cargo_.size()) {
-        return nullptr;
-    }
-    return cargo_[index];
+
+
```
Comment (mariuszlisowski) [inline shm/src/ship.cpp:52]: two newlines redundant

File: shm/src/ship.cpp [hunk 4]
```diff
@@ -41,18 +41,37 @@ size_t Ship::getSpeed() const     { return speed_; }
...
-                                      cargo_.end(),
-                                      0,
-                                      [](size_t sum, const auto& cargo) {
-                                          return sum += cargo->getAmount();
-                                      });
-    return capacity_ - occupied;
+    Common::getAvailableSpace(capacity_, cargo_);
+}
+
+bool Ship::addCargo(Cargo* cargo) {
```
Comment (mariuszlisowski) [inline shm/src/ship.cpp]: void instead of bool

File: shm/src/ship.cpp [hunk 5]
```diff
@@ -41,18 +41,37 @@ size_t Ship::getSpeed() const     { return speed_; }
...
+            cargo_.push_back(cargo);
+            //If there is already relevant type of cargo
+        } else {
+            size_t tmpAmount = ptrCargo->getAmount() + cargo->getAmount();
+            ptrCargo->setAmount(tmpAmount);
+        }
+    }
+}
+
+bool Ship::removeCargo(Cargo* cargo) {
```
Comment (mariuszlisowski) [inline shm/src/ship.cpp]: void instead of bool

File: shm/inc/ship.hpp [hunk 6]
```diff
@@ -24,15 +27,20 @@ class Ship {
     std::string getName() const;
     size_t getId() const;
     
-    std::shared_ptr<Cargo> getCargo(const size_t index);
+    Cargo* getCargo(const std::string& name);
     size_t getAvailableSpace() const;
 
+    void printCargo();
+    bool addCargo(Cargo* cargo);
```
Comment (mariuszlisowski) [inline shm/inc/ship.hpp]: void instead of bool

File: shm/inc/ship.hpp [hunk 7]
```diff
@@ -24,15 +27,20 @@ class Ship {
...
+    bool removeCargo(Cargo* cargo);
```
Comment (mariuszlisowski) [inline shm/inc/ship.hpp]: void instead of bool

File: shm/inc/player.hpp [hunk 8]
```diff
@@ -10,13 +10,19 @@ class Player {
...
-    size_t getAvailableSpace() const { return availableSpace_; }
+    size_t getAvailableSpace() const { return this->ship_->getAvailableSpace(); }
     size_t getSpeed() const { return this->ship_->getSpeed(); }
     std::shared_ptr<Cargo> getCargo(size_t index) const;
+    Cargo* getCargo(const std::string& name) const;
+
+    void setMoney(const size_t amount) { money_ = amount; }
+    void setAvaiableSpace(const size_t amount) { availableSpace_ = amount; }
+    void printShipCargo();
+    bool addCargo(Cargo* cargo) { return ship_->addCargo(cargo); }
```
Comment (mariuszlisowski) [inline shm/inc/player.hpp]: void instead of bool and no return

File: shm/inc/player.hpp [hunk 9]
```diff
@@ -10,13 +10,19 @@ class Player {
...
+    bool removeCargo(Cargo* cargo) { return ship_->removeCargo(cargo); }
```
Comment (mariuszlisowski) [inline shm/inc/player.hpp]: void instead of bool and no return

File: shm/inc/store.hpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    Response buy(Cargo* cargo, size_t amount, Player* player);
+    Response sell(Cargo* cargo, size_t amount, Player* player);
+    size_t getAvaiableSpace() const;
+    size_t getAvaiableFunds() const { return avaiableFunds_; };
+    void setCapasity(const size_t amount) { capacity_ = amount; }
+    void setAvaiableFunds(const size_t amount) { avaiableFunds_ = amount; }
+
+    bool addCargo(Cargo* cargo);
+    // bool addCargo(const std::string& name, size_t amount, size_t basePrice);  //overloaded
```
Comment (mariuszlisowski) [inline shm/inc/store.hpp]: void instead of bool

File: shm/src/store.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    return Response::invalid_cargo_name;
+}
+
+/*public*/
+//-----------------------------------------------------------------------------------
+// <summary> This method allows to add cargo to the store / updated the cargo vec
+//-----------------------------------------------------------------------------------
+bool Store::addCargo(Cargo* cargo) {
```
Comment (mariuszlisowski) [inline shm/src/store.cpp:81]: void instead of bool

**Review by mariuszlisowski — CHANGES_REQUESTED** [review]
Loads of code! Not an easy task. Congrats!

File: shm/inc/common.hpp [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+#include "cargo.hpp"
+
+class Common {
+public:
+    static void printCargo(const std::vector<Cargo*>& vecOfCargo);
+    static Cargo* getCargo(const std::string& name,
+                           const std::vector<Cargo*>& vecOfCargo);
+    static size_t getAvailableSpace(size_t capacity,
+                                    const std::vector<Cargo*>& vecOfCargo);
+};
```
Comment (mariuszlisowski) [inline shm/inc/common.hpp]: newline at the end

File: shm/src/store.cpp [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    } else if (mode == SELL) {
+        size_t tmpAmount = player->getCargo(cargo->getName())->getAmount() - amount;
+        player->getCargo(cargo->getName())->setAmount(tmpAmount);
+
+        if (this->getCargo(cargo->getName()) != nullptr) {
+            size_t tmpAmount = this->getCargo(cargo->getName())->getAmount() + amount;
+            this->getCargo(cargo->getName())->setAmount(tmpAmount);
+
+        } else {
+            std::cout << "You cant sell this type of good i stroe." << std::endl;
```
Comment (mariuszlisowski) [inline shm/src/store.cpp]: Message typo error

File: shm/src/ship.cpp [hunk 14]
```diff
@@ -41,18 +41,37 @@ size_t Ship::getSpeed() const     { return speed_; }
...
+    auto it = std::find_if(cargo_.begin(),
+                           cargo_.end(),
+                           [=](Cargo* el) {
+                               return el->getName() == cargo->getName();
+                           });
+    cargo_.erase(it);
 }
```
Comment (mariuszlisowski) [inline shm/src/ship.cpp]: printCargo() declared in ship.hpp but not implemented. Should be as follows:

void Ship::printCargo() {
    Common::printCargo(cargo_);
}

File: shm/src/store.cpp [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+
+      player->setMoney(player->getMoney() - totalPrice);
+      this->setAvaiableFunds(this->getAvaiableFunds() + totalPrice);
+      this->updateCargo(cargo, amount, updateMode::BUY, player);
+
+      return Response::done;
+    }
+
+    return Response::invalid_cargo_name;
+  }
```
Comment (mariuszlisowski) [inline shm/src/store.cpp]: error: control reaches end of non-void function
Solution: pull the return statement out of the if condition

File: shm/src/store.cpp [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+      // need update player cargo
+
+      player->setMoney(player->getMoney() + totalPrice);
+      this->setAvaiableFunds(this->getAvaiableFunds() - totalPrice);
+      this->updateCargo(cargo, amount, updateMode::SELL, player);
+      return Response::done;
+    }
+
+    return Response::invalid_cargo_name;
+  }
```
Comment (mariuszlisowski) [inline shm/src/store.cpp]: error: control reaches end of non-void function
Solution: pull the return statement out of the if condition

File: shm/src/cargo.cpp [hunk 17]
```diff
@@ -5,3 +5,9 @@ Cargo::Cargo(const std::string& name, size_t amount, size_t basePrice)
     , amount_(amount)
     , basePrice_(basePrice)
 {}
+
+Cargo::Cargo(const Cargo& cargo){
+    name_ = cargo.name_;
+    amount_ = cargo.amount_;
+    basePrice_ = cargo.basePrice_;
+}
```
Comment (mariuszlisowski) [inline shm/src/cargo.cpp]: newline at the end

File: shm/inc/ship.hpp [hunk 18]
```diff
@@ -24,15 +27,22 @@ class Ship {
     std::string getName() const;
     size_t getId() const;
     
-    std::shared_ptr<Cargo> getCargo(const size_t index);
+    Cargo* getCargo(const std::string& name);
     size_t getAvailableSpace() const;
 
+    void printCargo() {Common::printCargo(cargo_);}
```
Comment (mariuszlisowski) [inline shm/inc/ship.hpp]: Definition should be in source file

File: shm/inc/ship.hpp [hunk 19]
```diff
@@ -24,15 +27,22 @@ class Ship {
...
+
 private:
     size_t capacity_;
     size_t maxCrew_;
     size_t crew_;
     size_t speed_;
     std::string name_;
     const size_t id_;
-    std::vector<std::shared_ptr<Cargo>> cargo_;
+    // std::vecto/r<std::shared_ptr<Cargo>> cargo_;
```
Comment (mariuszlisowski) [inline shm/inc/ship.hpp]: Unused code (not needed)

File: shm/src/ship.cpp [hunk 20]
```diff
@@ -2,57 +2,97 @@
...
-    return capacity_ - occupied;
+  return Common::getAvailableSpace(capacity_, cargo_);
+}
+
+bool Ship::addCargo(Cargo* cargo) {
+  bool success = false;
+  if (this->getAvailableSpace() >= cargo->getAmount()) {
+    auto ptrCargo = this->getCargo(cargo->getName());
+
+    // If there is not such cargo type yet
```
Comment (mariuszlisowski) [inline shm/src/ship.cpp]: Comment redundant

File: shm/src/ship.cpp [hunk 21]
```diff
@@ -2,57 +2,97 @@
...
+    if (ptrCargo == nullptr) {
+      cargo_.push_back(cargo);
+      // If there is already relevant type of cargo
```
Comment (mariuszlisowski) [inline shm/src/ship.cpp]: Comment redundant

File: shm/inc/store.hpp [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+    ~Store() = default;
+
+    Response buy(Cargo* cargo, size_t amount, Player* player);
+    Response sell(Cargo* cargo, size_t amount, Player* player);
+    size_t getAvaiableSpace() const;
+    size_t getAvaiableFunds() const { return avaiableFunds_; };
+    void setCapasity(const size_t amount) { capacity_ = amount; }
+    void setAvaiableFunds(const size_t amount) { avaiableFunds_ = amount; }
+
+    bool addCargo(Cargo* cargo);
```
Comment (mariuszlisowski) [inline shm/inc/store.hpp]: Method repeated (DRY practice violation). Previous definition in ship.cpp. Please check if the method is really needed here.

**Review by mariuszlisowski — CHANGES_REQUESTED** [review]
Please check formatting in ship.cpp, store.cpp (indentation 2 spaces only).
Comments are redundant, imho.

File: shm/src/store.cpp [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+        return Response::lack_of_cargo;
+
+      if (this->getAvaiableSpace() <
+          player->getCargo(cargo->getName())->getAmount())
+        return Response::lack_of_space;
+
+      player->setMoney(player->getMoney() + totalPrice);
+      // need update player cargo
+
+      player->setMoney(player->getMoney() + totalPrice);
```
Comment (mariuszlisowski) [inline shm/src/store.cpp]: Line doubled

**Review by mariuszlisowski — APPROVED** [review]

File: shm/inc/player.hpp [hunk 24]
```diff
@@ -11,14 +11,25 @@ class Player : public Delegate {
 
     std::unique_ptr<Ship> getShip() { return std::move(ship_); }
     size_t getMoney() const { return money_; }
-    size_t getAvailableSpace() const { return availableSpace_; }
+    size_t getAvailableSpace() const { return this->ship_->getAvailableSpace(); }
```
Comment (m-kop) [inline shm/inc/player.hpp]: remove this

File: shm/src/ship.cpp [hunk 25]
```diff
@@ -23,18 +23,18 @@ Ship::~Ship() {
 }
 
 Ship& Ship::operator-=(const size_t crew) {
-    if (crew <= crew_) {
-        crew_ -= crew;
-        return *this;
-    }
```
Comment (m-kop) [inline shm/src/ship.cpp]: please do not change these lines

File: shm/src/ship.cpp [hunk 26]
```diff
@@ -23,18 +23,18 @@ Ship::~Ship() {
...
+    crew_ -= crew;
     return *this;
+  }
+  return *this;
 }
 Ship& Ship::operator+=(const size_t crew) {
-    if (crew_ + crew <= maxCrew_) {
-        crew_ += crew;
-        return *this;
-    }
```
Comment (m-kop) [inline shm/src/ship.cpp]: ditto

File: shm/src/ship.cpp (hunk 4 above)
Comment (m-kop) [reply shm/src/ship.cpp]: why? bool makes sense to me

File: shm/inc/store.hpp [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+    Store();
+    Store(size_t capacity, size_t avaiableFunds);
+    ~Store() = default;
+
+    Response buy(Cargo* cargo, size_t amount, Player* player);
+    Response sell(Cargo* cargo, size_t amount, Player* player);
+    void addCargo(Cargo* cargo);
+    size_t getAvaiableSpace() const;
+    size_t getAvaiableFunds() const { return avaiableFunds_; };
+    void setCapasity(const size_t amount) { capacity_ = amount; }
```
Comment (m-kop) [inline shm/inc/store.hpp]: typo -> Capacity

File: shm/inc/fruit.hpp [hunk 28]
```diff
@@ -25,6 +25,8 @@ class Fruit : public Cargo {
     size_t getTimeToSpoil() const;
     size_t getTimeToSpoilLeft() const;
 
+    Cargo* clone() override { return new Fruit( *this);}
```
Comment (m-kop) [inline shm/inc/fruit.hpp]: this memory is not deleted at all!!!

File: shm/inc/ship.hpp [hunk 29]
```diff
@@ -30,17 +33,23 @@ class Ship : public Observer {
     std::string getName() const;
     size_t getId() const;
     
-    std::shared_ptr<Cargo> getCargo(const size_t index);
+    Cargo* getCargo(const std::string& name);
```
Comment (m-kop) [inline shm/inc/ship.hpp:34]: I would keep shared_ptr and change the other calls

**Review by m-kop — CHANGES_REQUESTED** [review]

**Pull request merged**